sklearn/metrics/cluster/_dbcv_internals.py

```diff
@@ -0,0 +1,311 @@
+        metrics, such as minkowski, Mahanalobis etc.
+
+    Returns
+    -------
+    The 'density separation' between the clusters specified by
+    `cluster_id1` and `cluster_id2`.
+
+    References
+    ----------
+    Moulavi, D., Jaskowiak, P.A., Campello, R.J., Zimek, A. and Sander, J.,
+    2014. Density-Based Clustering Validation. In SDM (pp. 839-847).
+    """
+    if metric == "precomputed":
+        sub_select = X[labels == cluster_id1, :][:, labels == cluster_id2]
+        distance_matrix = sub_select[internal_nodes1, :][:, internal_nodes2]
+    else:
+        cluster1 = X[labels == cluster_id1][internal_nodes1]
+        cluster2 = X[labels == cluster_id2][internal_nodes2]
+        distance_matrix = cdist(cluster1, cluster2, metric, **kwd_args)
+
+    if no_coredist:
+        return distance_matrix.min()
+
+    else:
+        core_dist_matrix1 = np.tile(
+            core_distances1[internal_nodes1], (distance_matrix.shape[1], 1)
+        ).T
+        core_dist_matrix2 = np.tile(
+            core_distances2[internal_nodes2], (distance_matrix.shape[0], 1)
+        )
+
+        mr_dist_matrix = np.dstack(
+            [distance_matrix, core_dist_matrix1, core_dist_matrix2]
+        ).max(axis=-1)
+
+        return mr_dist_matrix.min()
+
+
+def internal_minimum_spanning_tree(mr_distances):
+    """
+    Compute the 'internal' minimum spanning tree given a matrix of mutual
+    reachability distances. Given a minimum spanning tree the 'internal'
+    graph is the subgraph induced by vertices of degree greater than one.
+
+    Parameters
+    ----------
+    mr_distances : array (cluster_size, cluster_size)
+        The pairwise mutual reachability distances, inferred to be the edge
+        weights of a complete graph. Since MSTs are computed per cluster
+        this is the all-points-mutual-reacability for points within a single
+        cluster.
+
+    Returns
+    -------
+    internal_nodes : array
+        An array listing the indices of the internal nodes of the MST
+
+    internal_edges : array (?, 3)
+        An array of internal edges in weighted edge list format; that is
+        an edge is an array of length three listing the two vertices
+        forming the edge and weight of the edge.
+
+    References
+    ----------
+    Moulavi, D., Jaskowiak, P.A., Campello, R.J., Zimek, A. and Sander, J.,
+    2014. Density-Based Clustering Validation. In SDM (pp. 839-847).
+    """
+    from ...cluster._hierarchical_fast import mst_linkage_core
+
+    single_linkage_data = mst_linkage_core(mr_distances, None, precomputed=True)
+    min_span_tree = single_linkage_data.copy()
+    for index, row in enumerate(min_span_tree[1:], 1):
+        candidates = np.where(np.isclose(mr_distances[int(row[1])], row[2]))[0]
+        candidates = np.intersect1d(
+            candidates, single_linkage_data[:index, :2].astype(int)
+        )
+        candidates = candidates[candidates != row[1]]
+        assert len(candidates) > 0
```

Comment:
Please do use assert only in tests. Here, raise a ValueError instead.

Reply:
Good point, I'll modify as suggested